### PR TITLE
Handle Delta Lake vacuum file system error

### PR DESCRIPTION
**Pull request by nevillelyh**
## Description

Error when deleting S3 files should be catagorized as file system error for accounting purposes.

Stack trace:

```
{
  "type": "io.trino.spi.TrinoException",
  "message": "Failure when vacuuming <table> with retention 7d: java.io.IOException: Exception while batch deleting paths",
  "cause": {
    "type": "java.lang.RuntimeException",
    "message": "Failure when vacuuming <table> with retention 7d: java.io.IOException: Exception while batch deleting paths",
    "cause": {
      "type": "java.io.IOException",
      "message": "Exception while batch deleting paths",
      "cause": {
        "type": "com.amazonaws.services.s3.model.MultiObjectDeleteException",
        "message": "One or more objects could not be deleted (Service: null; Status Code: 200; Error Code: null; Request ID: <request_id>; S3 Extended Request ID: <request_id> ; Proxy: null)",
        "suppressed": [],
        "stack": [
          "com.amazonaws.services.s3.AmazonS3Client.deleteObjects(AmazonS3Client.java:2388)",
          "io.trino.hdfs.s3.TrinoS3FileSystem.deletePaths(TrinoS3FileSystem.java:819)",
          "io.trino.hdfs.s3.TrinoS3FileSystem.deleteFiles(TrinoS3FileSystem.java:799)",
          "io.trino.filesystem.hdfs.HdfsFileSystem.lambda$deleteFiles$2(HdfsFileSystem.java:139)",
          "io.trino.hdfs.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:25)",
          "io.trino.hdfs.HdfsEnvironment.doAs(HdfsEnvironment.java:134)",
          "io.trino.filesystem.hdfs.HdfsFileSystem.deleteFiles(HdfsFileSystem.java:137)",
          "io.trino.filesystem.manager.SwitchingFileSystem.deleteFiles(SwitchingFileSystem.java:88)",
          "io.trino.filesystem.tracing.TracingFileSystem.lambda$deleteFiles$1(TracingFileSystem.java:79)",
          "io.trino.filesystem.tracing.Tracing.lambda$withTracing$1(Tracing.java:38)",
          "io.trino.filesystem.tracing.Tracing.withTracing(Tracing.java:47)",
          "io.trino.filesystem.tracing.Tracing.withTracing(Tracing.java:37)",
          "io.trino.filesystem.tracing.TracingFileSystem.deleteFiles(TracingFileSystem.java:79)",
          "io.trino.plugin.deltalake.procedure.VacuumProcedure.doVacuum(VacuumProcedure.java:296)",
          "io.trino.plugin.deltalake.procedure.VacuumProcedure.vacuum(VacuumProcedure.java:135)",
          "io.trino.plugin.objectstore.MethodHandles$Translate.invoke(MethodHandles.java:83)",
          "java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)",
          "java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:758)",
          "io.trino.execution.CallTask.execute(CallTask.java:210)",
          "io.trino.execution.CallTask.execute(CallTask.java:70)",
...
```

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java [hunk 1]
```diff
@@ -137,6 +138,9 @@ public void vacuum(
         catch (TrinoException e) {
             throw e;
         }
+        catch (IOException e) {
+            throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, format("Failure when vacuuming %s.%s with retention %s: %s", schema, table, retention, e), e);
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java:142]: This approach differs from a comment in the following `Exception` catch block. I think we should find a different solution.

Comment (nevillelyh) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java:142]: Any suggestion? We'd like to wrap these "known" errors as Trino `EXTERNAL_ERROR` for accounting purposes. This branch does that while leaving other unforeseen exceptions untouched, which will eventually be wrapped in a Trino `INTERNAL_ERROR`.

**Comment by mosabua** [comment]
fyi @findinpath

**Comment by findinpath** [comment]
@ebyhr is there anything that we should do extra here?
I see you did in the meantime 7a863b30c1448024feb92bbd6119943d977f80b5 to expose more information about the error.

Is **storage** an _internal_ part of Trino  (legalizing `INTERNAL_ERROR`) or rather an _external_ part?

cc @electrum @mayankvadariya @pajaks

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java [hunk 2]
```diff
@@ -137,6 +138,9 @@ public void vacuum(
...
+        }
         catch (Exception e) {
```
Comment (electrum) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java:144]: This should be changed to `catch (RuntimeException)` as there are no other checked exceptions to catch.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java (hunk 1 above)
Comment (electrum) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/VacuumProcedure.java:142]: This seems reasonable to me, and matches what we do in other places. File system errors should be categorized as `EXTERNAL_ERROR`.

@ebyhr did you have something different in mind?

**Pull request merged**